### ISSUE TITLE
[Feature Request] Double DQN

### ISSUE THREAD

**Issue by NickLucche** [DLR-RM/stable-baselines3#487]
### 🚀 Feature

Add double variant of the dqn algorithm.

### Motivation

It's in the roadmap https://github.com/DLR-RM/stable-baselines3/issues/1.

### Pitch

I suggest we go from:
```
with th.no_grad():
      # Compute the next Q-values using the target network
      next_q_values = self.q_net_target(replay_data.next_observations)
      # Follow greedy policy: use the one with the highest value
      next_q_values, _ = next_q_values.max(dim=1)
```
to:
```
with th.no_grad():
      # Compute the next Q-values using the target network
      next_q_values = self.q_net_target(replay_data.next_observations)
      if self.double_dqn:
          # use current model to select the action with maximal q value
          max_actions = th.argmax(self.q_net(replay_data.next_observations), dim=1)
          # evaluate q value of that action using fixed target network
          next_q_values = th.gather(next_q_values, dim=1, index=max_actions.unsqueeze(-1))
      else:
          # Follow greedy policy: use the one with the highest value
          next_q_values, _ = next_q_values.max(dim=1)
```
with `double_dqn` as additional flag to be passed to DQN init.

### Checklist

- [ x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
Yup indeed this is one of the core enhancements we would like to see!

@araffin just checking with you: is bloating DQN code here slightly ok (with a new parameter), or should this go to contrib?

@NickLucche If you wish to work on this PR, note the we would like to see matching results with reference implementations (e.g. same results as in the original paper or in a well-established codebase). One or two Atari games should be enough.

**Comment by NickLucche** [comment]
Sure thing, I'll make sure to setup the experiments in a similar way.
I'll do some searching starting from the source paper.

**Comment by Miffyli** [comment]
I highly recommend using [zoo](https://github.com/DLR-RM/rl-baselines3-zoo/) for doing that btw, since it already has known-to-work preprocessing for Atari :). It takes a moment to get into, but will definitely pay itself off with these experiments (there are many tiny details that need taking of).

**Comment by araffin** [comment]
>@araffin just checking with you: is bloating DQN code here slightly ok (with a new parameter), or should this go to contrib?

I think double DQN is ok as it is only a 3 lines change.
However, for prioritized replay and the rest of DQN extensions, I would prefer to have a RAINBOW implementation with everything in it. So we keep DQN implementation clean but have the full-bloated RAINBOW too.

**Comment by NickLucche** [comment]
@Miffyli Thanks for the tip.

It may take a while to run the experiments for the same number of timesteps as in "Human-level control through deep reinforcement learning" tho.
e.g I'm getting a decent improvement after ~5.5mln steps in terms of mean reward ( currently 5.92e+03 vs 1.41e+04) on Atlantis-v0.
Can I cut short on the number of steps or should I match the experiments in that too?  What kind of benchmarks do you expect as a result?

**Comment by araffin** [comment]
> Can I cut short on the number of steps or should I match the experiments in that too? What kind of benchmarks do you expect as a result?

I could help with that if needed but basically you can check what we did with QR-DQN: https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/13

So Breakout + Pong using 10e6 training steps and 3 seeds (40e6 frames because of frame skip) + some benchmark on simple gym task.

**Comment by NickLucche** [comment]
Thanks a lot for your help, it is most appreciated!
I'd love to help out with this small contribution, so I'll try to mimic what has been done in that issue, if that's okay with you.

I was using a script very similar to this for testing so far https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/13#issuecomment-742913092 as I'm yet not super-familiar with zoo.
Is there anything which can help out speeding the benchmarking process from which I should start/bootstrap from..?

**Comment by araffin** [comment]
>I was using a script very similar to this for testing so far Stable-Baselines-Team/stable-baselines3-contrib#13 (comment) as I'm yet not super-familiar with zoo.

RL Zoo is quite easy to use, if you want to replicate results and plot them, we have a section for that in the documentation ;)
https://stable-baselines3.readthedocs.io/en/master/modules/dqn.html#how-to-replicate-the-results

**Comment by NickLucche** [comment]
Sorry for the long inactivity. 
I managed to run a few experiments with the proposed change on pong+breakout, I'll leave here the training curves although I can't notice much of a difference (at least that's on par with original papers findings).

## Vanilla DQN Pong
![pong_Training_Episodic_Reward](https://user-images.githubusercontent.com/10706289/133688000-e0da9688-90e3-43b7-8f7f-a02e685bd20b.png)
## Vanilla DQN Breakout
![breakout_Training_Episodic_Reward](https://user-images.githubusercontent.com/10706289/133688133-47d2c172-83fb-4828-a69f-592d0702d8fe.png)
## Double DQN Pong
![double_pong_Training_Episodic_Reward](https://user-images.githubusercontent.com/10706289/133688149-5378d6ed-171c-446a-b1e1-3338918f4899.png)
## Double DQN Breakout
![breakout_double_Training_Episodic_Reward](https://user-images.githubusercontent.com/10706289/133688163-2c40e3ed-915f-4f3f-a2f1-12ada88dc0c0.png)

**Comment by araffin** [comment]
thanks for the update, best would be to have both dqn and double dqn in the same plot (i can help you with the command if needed).
Also, you should probably try on Atari games where double dqn is known to make a difference (in the paper, they give some examples, i think Asterix is one of them)

**Comment by NickLucche** [comment]
Oh yeah I'd appreciate that, I've used the `plot_train` script from rl-zoo to be more standard but I can't figure out how to put the results in the same plot using that.

Sure thing, I'll make sure to test it on a few more interesting use-cases.

**Comment by araffin** [comment]
> Oh yeah I'd appreciate that, I've used the plot_train script from rl-zoo to be more standard but I can't figure out how to put the results in the same plot using that.

cf. https://stable-baselines3.readthedocs.io/en/master/modules/dqn.html#how-to-replicate-the-results

you need to plot the evaluations, for instance comparing dqn and ppo on Pong and Breakout:
```
python scripts/all_plots.py -a dqn ppo -e Pong Breakout -f rl-trained-agents/
```
(you can also specify multiple folders)

**Comment by araffin** [comment]
closing in favor of #622
